bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/grpc/RefreshingOAuth2CredentialsInterceptor.java

```diff
@@ -0,0 +1,216 @@
+      this.exception = null;
+      this.header = cachedHeader;
+      this.staleTime = staleTime;
+      this.expiresTime = expiresTime;
+    }
+
+    public TokenCache(IOException exception) {
+      super();
+      this.exception = exception;
+      this.header = null;
+      this.staleTime = -1;
+      this.expiresTime = -1;
+    }
+
+    public CacheState getCacheState() {
+      long now = System.currentTimeMillis();
+      if (now > expiresTime) {
+        return CacheState.Expired;
+      } else if (now > staleTime) {
+        return CacheState.Stale;
+      } else {
+        return CacheState.Good;
+      }
+    }
+  }
+  
+  /**
+   * This specifies how far in advance of a token expiration do we consider the token stale.
+   */
+  private static final int TOKEN_STALENESS_MS = 60 * 1000;
+  private static final int TOKEN_EXPIRES_MS = 1 * 1000;
+
+
+  private static final Metadata.Key<String> AUTHORIZATION_HEADER_KEY =
+      Metadata.Key.of("Authorization", Metadata.ASCII_STRING_MARSHALLER);
+
+
+  private final AtomicReference<TokenCache> tokenCache = new AtomicReference<>();
+  private final AtomicBoolean isRefreshing = new AtomicBoolean(false);
+  private final ExecutorService executor;
+  private final OAuth2Credentials credentials;
+
+  @VisibleForTesting
+  // Allow clock to be overridden by test code.
+  Clock clock = Clock.SYSTEM;
+
+  public RefreshingOAuth2CredentialsInterceptor(ExecutorService scheduler,
+      OAuth2Credentials credentials) {
+    this.executor = Preconditions.checkNotNull(scheduler);
+    this.credentials = Preconditions.checkNotNull(credentials);
+  }
+
+  @Override
+  public <ReqT, RespT> Call<ReqT, RespT> interceptCall(MethodDescriptor<ReqT, RespT> method,
+                                                       Channel next) {
+    // TODO(ejona86): If the call fails for Auth reasons, this does not properly propagate info that
+    // would be in WWW-Authenticate, because it does not yet have access to the header.
+    return new CheckedForwardingCall<ReqT, RespT>(next.newCall(method)) {
+      @Override
+      protected void checkedStart(Listener<RespT> responseListener, Metadata.Headers headers)
+          throws Exception {
+        headers.put(AUTHORIZATION_HEADER_KEY, getHeader());
+        delegate().start(responseListener, headers);
+      }
+    };
+  }
+
+  public void asyncRefresh() {
+    if (!isRefreshing.get()) {
+      executor.execute(new Runnable() {
+        @Override
+        public void run() {
+          boolean requiresRefresh = false;
+          synchronized (isRefreshing) {
+            if (!isRefreshing.get()) {
+              isRefreshing.set(true);
+              requiresRefresh = true;
+            }
+          }
+          if (requiresRefresh) {
+            doRefresh();
+          }
+        }
+      });
+    }
+  }
+
+  /**
+   * Get the oauth2 AccessToken.
+   */
+  public String getHeader() throws IOException {
+    TokenCache tokenCache = getCachedHeader();
```

Comment:
I think that if token.exception is not null, getCachedHeader will always raise an exception, regardless of it being stale or expired, and it will never be refreshed.